### PR TITLE
Fix pull parameter parsing for compat `/build` endpoint

### PR DESCRIPTION
**Pull request by matejvasek**
# Changes

Fixing Docker client incompatibility:
```
# Broken compatibility with the docker CLI
docker build . -t alpine-1000 --pull
Sending build context to Docker daemon  2.048kB
Error response from daemon: failed to parse query parameter 'pull': "1": invalid pull policy: "1"
```

* Revert https://github.com/containers/podman/pull/18583 since it breaks compatibility with standard docker client.
Previously it was believed the `pull` parameter is pull-policy since REST API doc says it is `string`, not `bool`. However it fact it is a boolean. Confusingly enough Docker indeed accept values like `always` or `never` but these values are interpreted as `true` by Docker. As matter of the fact it appears that there is no such a thing as `pull-policy` in Docker.
* Use custom boolean query param deserializer for all compat endpoints to get closer to Docker's way.
This means anything but `""`, `"0"`, `"no"`, `"false"`, `"none"` (ignoring case) is considered to be true.

More context: https://github.com/containers/podman/issues/17778#issuecomment-1673931925

```release-note
Correct parsing of the pull query parameter for the compat /build endpoint.
```

### CODE REVIEW

**Comment by matejvasek** [comment]
I do not insist on [the second commit](https://github.com/containers/podman/pull/19591/commits/eb56909968d31d11e24a384cc4e3d475d73b804c) but we definitely must revert https://github.com/containers/podman/pull/18583.

**Comment by matejvasek** [comment]
Or you do not like proliferation of `CompatDecoderKey` we can use that specific decoder only for the endpoint in question (at least for now).

**Comment by matejvasek** [comment]
wrt `DCO`: do I really need to sign revert commit?

**Review by flouthoc — APPROVED** [review]
LGTM

Its weird though to drift away from docs here: https://docs.docker.com/engine/api/v1.42/#tag/Image/operation/ImageBuild ( but since your investigation points that there might be an issue with the docker documentation itself, I am convinced )  lets wait for other maintainers.

This is a breaking change even if this is reverting to an old behavior.

**Comment by vrothberg** [comment]
Please add rather verbose explanations on _why_ we are reverting the commit.  It has already been released.

**Comment by rhatdan** [comment]
Is there a test to make sure the libpod entry point still works with the string passing?

**Comment by matejvasek** [comment]
@rhatdan I don't know, but I think I have not touched libpod part at all.

**Comment by matejvasek** [comment]
One questionable thing about this PR is changing boolean parser for all query params of all compat endpoints.

**Comment by matejvasek** [comment]
> One questionable thing about this PR is changing boolean parser for all query params of all compat endpoints.

But it changes it in **mostly** compatible way. Strings "true", "1" are still `true` and "false", "0" are still `false`. However string "f" will be now interpreted as `true`, previously it was `false`. But this is how Docker converts string to booleans.

**Comment by matejvasek** [comment]
@vrothberg @Luap99 I updated description.

**Comment by matejvasek** [comment]
@rhatdan wdyt about https://github.com/containers/podman/pull/19591#issuecomment-1674821325 ?

**Comment by Luap99** [comment]
The problem is a lot of the compat endpoint are also used for the libpod endpoints. And I Really do not want such ugly bool parsing for libpod endpoints, sending using `opt=blah` should return a error to the caller and not assume true IMO. A single typo could be fatal otherwise.

Regardless I think for docker endpoints this is reasonable, so I guess if you make sure to guard the decoder step
`decoder := r.Context().Value(api.DecoderKey).(*schema.Decoder)` with `if utils.IsLibpodRequest(r)` and use the correct decoder based on that it should not cause any problem for any libpod endpoints.

**Comment by matejvasek** [comment]
>sending using `opt=blah` should return a error to the caller and not assume true IMO

@Luap99  Wholeheartedly agree, but tell that to `moby`.

**Comment by matejvasek** [comment]
Also as I said I am fine with just reverting to  previous state where presumably `strconv.ParseBool()` was used.

File: pkg/api/handlers/compat/images_build.go [hunk 1]
```diff
@@ -580,19 +579,8 @@ func BuildImage(w http.ResponseWriter, r *http.Request) {
 		pullPolicy = buildahDefine.PolicyMap[query.PullPolicy]
 	} else {
 		if _, found := r.URL.Query()["pull"]; found {
```
Comment (matejvasek) [inline pkg/api/handlers/compat/images_build.go:581]: @Luap99 Is this `found` guard needed? Why?

**Comment by matejvasek** [comment]
@Luap99 I fixed decoder discrimination for libpod and compat endpoints.

File: pkg/api/handlers/compat/images_build.go (hunk 1 above)
Comment (matejvasek) [reply pkg/api/handlers/compat/images_build.go:581]: I mean if the pull param is not present then quar.yPull is false anyway right?

Comment (Luap99) [reply pkg/api/handlers/compat/images_build.go:581]: Yes I don't think that is needed and should work without it.

**Review by Luap99 — COMMENTED** [review]
LGTM

**Comment by matejvasek** [comment]
Little bit more conservative fix https://github.com/containers/podman/pull/19601/files

**Comment by matejvasek** [comment]
@Luap99 @rhatdan  IMO it would be good to backport this into patch releases. If this PR seems to risky for backporting we can consider something more conservative like https://github.com/containers/podman/pull/19601/

**Comment by vrothberg** [comment]
> @vrothberg @Luap99 I updated description.

I don't see it in the commits.

**Comment by flouthoc** [comment]
@matejvasek I think PR description is updated but still it looks less verbose as compared to your investigation here https://github.com/containers/podman/issues/17778#issuecomment-1673931925, I'd vote for including this in commit body or alteast a link to it: https://github.com/containers/podman/issues/17778#issuecomment-1673931925

**Comment by matejvasek** [comment]
@vrothberg @flouthoc  I updated commit message.

**Comment by matejvasek** [comment]
@rhatdan will this be eligible for 4.6.2?

**Comment by Luap99** [comment]
I think backporting https://github.com/containers/podman/pull/19601 makes more sense IMO as this one has less risk involved, if you need the backport in RHEL you need to file a bugzilla as 4.6.1-rhel is already frozen for RHEL.

**Comment by matejvasek** [comment]
What will be time period between 4.6.1-rhel and 4.6.2-rhel ? @Luap99

**Comment by Luap99** [comment]
There will never be a 4.6.2 in RHEL, RHEL is frozen for each minor RHEL release. Then only approved backports are allowed to the rhel branches so you need to file bugzillas (or Jira bugs soon) for those in order to get this fixed in RHEL.
4.6.1 will go into RHEL 8.9/9.3 and then next version will be only in 6 months for 8.10/9.4 which by then will likely by podman 4.8 or later whatever we decide upstream.

**Comment by matejvasek** [comment]
https://github.com/containers/podman/pull/19631

**Comment by matejvasek** [comment]
@Luap99  https://issues.redhat.com or somewhere else?

**Comment by TomSweeneyRedHat** [comment]
@matejvasek I think this is what you're looking for: https://bugzilla.redhat.com/show_bug.cgi?id=2232127

**Comment by matejvasek** [comment]
@TomSweeneyRedHat yeah I created that issue after chat on Slack with @Luap99 .

**Pull request merged**